open-craft/sprints
# [BB-3174] Automate estimation session

This adds estimation session automation for the [Agile Poker](https://marketplace.atlassian.com/apps/700473/agile-poker-for-jira-planning-estimation) Jira app.

# Prerequisites
1. Set `.env` (you can find them [here](https://vault.opencraft.com:8200/ui/vault/secrets/secret/show/core/Sprints%20Development%20Envs)). There were some recent changes.
1. To `sprints/dashboard/tasks.py:update_estimation_session_task`, add the following below `for cell in get_cells(conn):`:
	```python
	if cell.board_id != 26:
	    continue
	```
1. To `sprints/dashboard/tasks.py:update_estimation_session_task`, add the following below `all_member_keys = list(current_member_keys | member_keys)`:
	```python
	print(f"Members of {cell.name}: {all_member_keys}")
	print(f"Unestimated tickets from {cell.name}: {[conn.issue(issue.id).key for issue in issues if issue.key.startswith(cell.key)]}")
	```
1. To `sprints/dashboard/tasks.py:move_estimates_to_tickets_task`, add the following below `for cell in get_cells(conn):`:
	```python
	if cell.board_id != 26:
	    continue
	```
1. To `sprints/dashboard/tasks.py:move_estimates_to_tickets_task`, add the following below `final_vote = get_poker_session_final_vote(votes, vote_values)`:
	```python
	print(f"The final vote for {conn.issue(issue).key} is {final_vote}.")
	```
1. To `sprints/dashboard/automation.py:ping_users_on_ticket`, add the following above `if not settings.DEBUG:`:
	```python
	print(message)
	```

# Testing instructions
1. Start the backend with `docker-compose -f local.yml up --build`.
1. Run migrations with `docker-compose -f local.yml exec django python manage.py migrate`.
1. Open Django shell with `docker-compose -f local.yml exec django python manage.py shell_plus` and run:
	```python
	from sprints.dashboard.tasks import *
	from sprints.dashboard.libs.jira import *
	
    conn = CustomJira(
        server=settings.JIRA_SERVER,
        basic_auth=(settings.JIRA_USERNAME, settings.JIRA_PASSWORD),
        options={
            'agile_rest_path': GreenHopperResource.AGILE_BASE_REST_PATH,
            'headers': {
                'Referer': 'https://tasks.opencraft.com/download/resources/com.spartez.jira.plugins.jiraplanningpoker/frontend/index.html',
            }
        }
    )
	```

1. To test creating estimation session, use the following snippet:
	```python
	session = conn.create_poker_session(
	    board_id=26,  # This is the Bebop one.
	    name=get_next_poker_session_name(conn),
	    issues=[],
	    participants=['viadanna'],
	    scrum_masters=[conn.myself()['key']],
	    send_invitations=False,
	)
	```
1. Visit the "Estimates" section in Jira and see that the session called "Sprint 241 & Stretch Goals" has been created. It should have only one button - "Estimate". Open the session and see that it does not work - the "Loading issues..." is there all the time. This is intended behavior, as explained in the `create_estimation_session_task`'s docstring.
1. Run the following snippet:
	```python
	session.update({'issuesIds': ['27146'], 'scrumMasters': ['crafty', 'viadanna']})
	```
1. Reload the page and check that `BB-1986` has been added there. Also, check that you can click on the `··· -> Configure session` button, as you are the scrum master of this session.
1. To test updating estimation session task, run the following snippet and check that the output looks reasonable:
	```python
	update_estimation_session_task()
	```
1. To verify retrieving session for a cell, run the following snippet and see that it does not raise any exception:
	```python
	assert len(conn.poker_sessions(26, state="OPEN", name=get_next_poker_session_name(conn))) == 1
	assert session.sessionId == conn.poker_sessions(26, state="OPEN", name=get_next_poker_session_name(conn))[0].sessionId
	```
1. Close the session with the following and verify that it's closed in Jira.
	```python
	conn.close_poker_session(session.sessionId, False)
	```
1. Run the following and check that it displays a message with a comment to me that it couldn't estimate the issue.
	```python
	move_estimates_to_tickets_task()
	```
1. Use `··· -> Restart session` in Jira to reopen the session (do not add any new tickets, just click `Next`). Vote for the ticket on the session page.
1. Run the following and check that it displays a message with the estimate you have submitted.
	```python
	move_estimates_to_tickets_task()
	```
1. Clean up by running the following:
	```python
	session.delete()
	```
